### PR TITLE
fix: emit events when node template fails to resolve

### PR DESCRIPTION
**Pull request by njtran**
Fixes #N/A 

**Description**
When a NodeTemplate fails to resolve in the cloudprovider code for a NotFoundError, Karpenter ignores it. Since this may happen very quickly and will not resolve itself, Karpenter now emits an event for this. 

Additionally, this change will not include cloudprovider.InstanceTypes() if they have no offerings. In the case of `karpenter-core`'s scheduling and deprovisioning logic, this means that Karpenter will consider less instance types in some cases when the zones fail to resolve. 

This is also the buddy change for https://github.com/aws/karpenter-core/pull/491

**How was this change tested?**
- make presubmit && snapshot tests.

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates 
- [ ] Yes, issue opened: # 
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by njtran — COMMENTED** [review]
/karpenter snapshot

**Comment by ellistarn** [comment]
Can we drive these events via status conditions?

File: go.mod [hunk 1]
```diff
@@ -107,3 +107,5 @@ require (
 	sigs.k8s.io/structured-merge-diff/v4 v4.2.3 // indirect
 	sigs.k8s.io/yaml v1.3.0 // indirect
 )
+
+replace github.com/aws/karpenter-core => github.com/njtran/karpenter-core v0.0.0-20230829002458-d8c86f22e7e6
```
Comment (jonathan-innis) [inline go.mod:111]: Do you just want to bump directly since this update should be merged now?

File: pkg/cloudprovider/cloudprovider.go [hunk 2]
```diff
@@ -40,6 +41,7 @@ import (
 	"knative.dev/pkg/logging"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
+	cloudproviderevents "github.com/aws/karpenter/pkg/cloudprovider/events"
```
Comment (jonathan-innis) [inline pkg/cloudprovider/cloudprovider.go:44]: I've changed my mind a bit about the utility of this being in a separate package. What do you think about collapsing the events package into the same package and suffixing the function call with `Event`?

File: pkg/cloudprovider/events/events.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/events"
+)
+
+func ProvisionerFailedToResolveNodeTemplate(provisioner *v1alpha5.Provisioner) events.Event {
+	return events.Event{
+		InvolvedObject: provisioner,
+		Type:           v1.EventTypeWarning,
+		Message:        "Failed to resolve AWSNodeTemplate",
```
Comment (jonathan-innis) [inline pkg/cloudprovider/events/events.go:28]:
```suggestion
		Message:        "Failed resolving AWSNodeTemplate",
```

nit: IMO, this sounds more natural

File: pkg/cloudprovider/events/events.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+		DedupeValues:   []string{provisioner.Name},
+	}
+}
+
+func MachineFailedToResolveNodeTemplate(machine *v1alpha5.Machine) events.Event {
+	return events.Event{
+		InvolvedObject: machine,
+		Type:           v1.EventTypeWarning,
+		Message:        "Failed to resolve AWSNodeTemplate",
```
Comment (jonathan-innis) [inline pkg/cloudprovider/events/events.go:37]:
```suggestion
		Message:        "Failed resolving AWSNodeTemplate",
```

nit: IMO, this sounds more natural

File: pkg/cloudprovider/events/events.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+		DedupeValues:   []string{provisioner.Name},
+	}
+}
+
+func MachineFailedToResolveNodeTemplate(machine *v1alpha5.Machine) events.Event {
+	return events.Event{
+		InvolvedObject: machine,
+		Type:           v1.EventTypeWarning,
```
Comment (jonathan-innis) [inline pkg/cloudprovider/events/events.go:36]: Is this a Warning? I think this should probably be an `Error` event?

File: pkg/cloudprovider/events/events.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/events"
+)
+
+func ProvisionerFailedToResolveNodeTemplate(provisioner *v1alpha5.Provisioner) events.Event {
+	return events.Event{
+		InvolvedObject: provisioner,
+		Type:           v1.EventTypeWarning,
```
Comment (jonathan-innis) [inline pkg/cloudprovider/events/events.go:27]: Is this a Warning? I think this should probably be an Error event?

File: pkg/providers/instancetype/suite_test.go [hunk 7]
```diff
@@ -1135,6 +1136,42 @@ var _ = Describe("Instance Types", func() {
 			}
 		})
 		It("shouldn't report more resources than are actually available on instances", func() {
+			awsEnv.EC2API.DescribeSubnetsOutput.Set(&ec2.DescribeSubnetsOutput{
```
Comment (jonathan-innis) [inline pkg/providers/instancetype/suite_test.go:1139]: Why did you need to add this?

**Review by jonathan-innis — COMMENTED** [review]
Nice work! Just a few thoughts

File: pkg/cloudprovider/cloudprovider.go (hunk 2 above)
Comment (njtran) [reply pkg/cloudprovider/cloudprovider.go:44]: I'd like to follow the semantic of how this is already done, and then opt to re-factor it in a separate PR, if that's alright with you. Why have you changed your mind?

File: pkg/cloudprovider/events/events.go (hunk 5 above)
Comment (njtran) [reply pkg/cloudprovider/events/events.go:36]: There are only Warning and Normal events, right? https://pkg.go.dev/k8s.io/api@v0.26.6/core/v1#:~:text=const%20(%0A%09//%20Information%20only%20and%20will%20not%20cause%20any%20problems%0A%09EventTypeNormal%20string%20%3D%20%22Normal%22%0A%09//%20These%20events%20are%20to%20warn%20that%20something%20might%20go%20wrong%0A%09EventTypeWarning%20string%20%3D%20%22Warning%22%0A)

File: pkg/providers/instancetype/suite_test.go (hunk 7 above)
Comment (njtran) [reply pkg/providers/instancetype/suite_test.go:1139]: Since we now reject and remove an instance types that have 0 offerings, we need to make sure the describe subnets returns a zone so that we can get an instance type that has at least one offering for the test below.

File: pkg/cloudprovider/cloudprovider.go (hunk 2 above)
Comment (jonathan-innis) [reply pkg/cloudprovider/cloudprovider.go:44]: > Why have you changed your mind

Creates extra packages for almost no reason. We've already started moving to this semantic in other PRs. Fair point though, I'm fine taking this separately.

File: pkg/cloudprovider/events/events.go [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+		DedupeValues:   []string{provisioner.Name},
+	}
+}
+
+func MachineFailedToResolveNodeTemplate(machine *v1alpha5.Machine) events.Event {
+	return events.Event{
+		InvolvedObject: machine,
+		Type:           v1.EventTypeWarning,
+		Message:        "Failed resolving AWSNodeTemplate",
+		DedupeValues:   []string{machine.Name},
```
Comment (jonathan-innis) [inline pkg/cloudprovider/events/events.go:38]: Should this be the name or the UID?

File: pkg/cloudprovider/events/events.go [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/events"
+)
+
+func ProvisionerFailedToResolveNodeTemplate(provisioner *v1alpha5.Provisioner) events.Event {
+	return events.Event{
+		InvolvedObject: provisioner,
+		Type:           v1.EventTypeWarning,
+		Message:        "Failed resolving AWSNodeTemplate",
+		DedupeValues:   []string{provisioner.Name},
```
Comment (jonathan-innis) [inline pkg/cloudprovider/events/events.go:29]: Should this be the name or the UID?

**Review by jonathan-innis — APPROVED** [review]
LGTM 🚀

**Review by njtran — COMMENTED** [review]
/karpenter snapshot

**Comment by njtran** [comment]
@ellistarn Definitely. I would see this as a new controller, but think this is a great opportunity for more observability into healthiness of various Karpenter components. Made an issue for this: https://github.com/aws/karpenter-core/issues/493
